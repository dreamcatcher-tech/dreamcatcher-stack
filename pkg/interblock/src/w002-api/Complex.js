import assert from 'assert-fast'
export default class Complex {
  #parent
  static create({ state, binary, network, actions, isLoading, wd, tree }) {
    return new Complex({ state, binary, network, actions, isLoading, wd, tree })
  }
  constructor({ state, binary, network, actions, isLoading, wd, tree }) {
    this.state = state || {}
    this.binary = binary
    this.network = network || []
    this.actions = actions || {}
    this.isLoading = isLoading || false
    this.wd = wd || '/'
    this.tree = tree
    if (tree === undefined) {
      this.tree = Complex.create({ ...this, tree: false })
    }
  }
  clone() {
    const next = new Complex(this)
    next.#parent = this.#parent
    return next
  }
  addAction(actions) {
    const next = this.clone()
    next.actions = { ...next.actions, ...actions }
    return next
  }
  setWd(wd) {
    assert.strictEqual(typeof wd, 'string')
    assert(wd.startsWith('/'))
    const next = this.clone()
    next.wd = wd
    return next
  }
  child(path) {
    assert.strictEqual(typeof path, 'string')
    assert(!path.startsWith('/'))
    assert(!path.startsWith('./'))
    assert(!path.startsWith('..'))
    const child = this.network.find(({ path: p }) => p === path)
    if (!child) {
      throw new Error(`child not found: ${path}`)
    }
    const { tree } = this
    const next = Complex.create({ ...child, tree })
    next.#parent = this
    return next
  }
  hasChild(path) {
    if (!path) {
      return false
    }
    assert.strictEqual(typeof path, 'string')
    assert(!path.startsWith('/'))
    assert(!path.startsWith('./'))
    assert(!path.startsWith('..'))
    return this.network.some(({ path: p }) => p === path)
  }
  rm(path) {
    assert(this.hasChild(path))
    const next = this.clone()
    next.network = next.network.filter(({ path: p }) => p !== path)
    return next
  }
  setNetwork(network) {
    assert(Array.isArray(network))
    const next = this.clone()
    next.network = network
    return next
  }
  setState(state) {
    assert.strictEqual(typeof state, 'object')
    const next = this.clone()
    next.state = state
    return next
  }
  parent() {
    return this.#parent
  }
}
