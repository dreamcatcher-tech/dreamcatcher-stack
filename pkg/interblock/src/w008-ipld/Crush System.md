# The Crush System

This collection of classes are designed to break a Pulse down into IPFS compatible chunks, for efficient hashing.

## Requirements

1. use the inheritance of IpldInterface as identifying feature of classes that get crushed
1. crush relies on an unbroken chain of crushable classes from the root down
   1. ie: cannot have a pojo that holds a crush class in it and expect it to work
1. arrays of crush classes follow these rules
1. when flagged, some children may be 'punched out' as CID links, and reinflated back in
1. any tree of classes can be reinflated from ipfs back into classes
1. ipfs blocks representing the different between the previous 'crush' process can be fetched then persisted in ipfs storage
1. be able to directly represent an IPFS primitive as a class, such as a CID - these are needed to signal to the ORM that it should stop rehydration
1. Be able to connect a set of IPLD schemas to the classes, to check them for correctness on crush and uncrush
   1. This check should be disablable for speed
