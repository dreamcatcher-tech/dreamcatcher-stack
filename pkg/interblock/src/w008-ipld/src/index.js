export { Binary } from './Binary'
export { RequestId } from './RequestId'
export { Request } from './Request'
export { RxRequest } from './RxRequest'
export { Reply } from './Reply'
export { RxReply } from './RxReply'
export { Address } from './Address'
export { PulseLink, HistoricalPulseLink } from './PulseLink'
export { TxQueue } from './TxQueue'
export { Tx } from './Tx'
export { RxQueue } from './RxQueue'
export { Rx } from './Rx'
export { Channel } from './Channel'
export { Config } from './Config'
export { PublicKey } from './PublicKey'
export { Validators } from './Validators'
export { ChannelsHamt } from './ChannelsHamt'
export { ChildrenHamt } from './ChildrenHamt'
export { UplinksHamt } from './UplinksHamt'
export { DownlinksHamt } from './DownlinksHamt'
export { SymlinksHamt } from './SymlinksHamt'
export { AddressesHamt } from './AddressesHamt'
export { Channels } from './Channels'
export { Transmissions } from './Transmissions'
export { Network } from './Network'
export { AsyncRequest } from './AsyncRequest'
export { AsyncTrail } from './AsyncTrail'
export { Pending } from './Pending'
export { State } from './State'
export { Timestamp } from './Timestamp'
export { Dmz } from './Dmz'
export { Provenance } from './Provenance'
export { Interpulse } from './Interpulse'
export { Pulse } from './Pulse'
export { Keypair } from './Keypair'
export { decode } from './IpldUtils'
export { IpldInterface } from './IpldInterface'
export { Hamt } from './Hamt'
