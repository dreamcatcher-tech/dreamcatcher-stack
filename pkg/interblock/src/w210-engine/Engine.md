## Overview

The purpose of the Engine is to execute and coordinate all possible chain behaviours that are valid. Behaviours include: merging, forking, consensus conflicts, verification, multi chain interaction, versioning, covenant upgrades

The Engine has 3 types of event that it can process:

1. Interpulse updates - new interpulse for a chain we validate has occured
1. Pool updates - where there is a new pool for a chain
1. Pulse updates - a new Pulse as been formed

The Engine operates in local mode by default, meaning that it uses javascripts single threaded nature to guarantee locks. The engine can be connected to a hints service, which does the job of sharing the work of this engine with other engines, and can sometimes trigger the 3 engine events out of turn because they were caused externally. Interpulse updates trigger a pool update which triggers a pulse update. Pool updates trigger a Pulse update. Pulse updates trigger a pool update as the pool need to be reconciled with the new Pulse.

Pools are used to propose the next pulse, with a Pool update being used to send round something with your signature on it

## Pierce

Pierce is the means by which external Requests and Replies enter into chainland. Pierce is considered a form of pool update.

## Subscriptions

Clients will normally subscribe to Pulse updates. Validators will always subscribe to Interpulse updates, and in a multivalidator group, will also subscribe to Pool updates.

## TODO

Engines operate in one of four roles relative to a given chain:

1. proposing
2. validating
3. witnessing
4. consuming

## Effects

Execution secondarily deals with the running of side effects. This is because the loading of a covenant is the same for block making as it is for effects.
