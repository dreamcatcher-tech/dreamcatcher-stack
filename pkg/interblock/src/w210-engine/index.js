export { Engine } from './src/Engine'
