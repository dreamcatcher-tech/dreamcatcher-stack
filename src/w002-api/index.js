const api = require('./src/api')
const hooks = require('./src/hooks')
const queries = require('./src/queries')
module.exports = { ...api, ...hooks, ...queries }
