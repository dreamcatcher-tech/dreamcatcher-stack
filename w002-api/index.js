const api = require('./src/api')
const promises = require('./src/promises')
module.exports = { ...api, ...promises }
