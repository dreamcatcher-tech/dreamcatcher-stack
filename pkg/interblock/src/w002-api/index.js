import assert from 'assert-fast'
import { Request } from '../w008-ipld'

export const interchain = (type, payload, to) => {
  const { interchain } = globalThis[Symbol.for('interblock:api:hook')]
  assert(interchain, `global hook detached`)
  return interchain(type, payload, to)
}
export const useState = (path = '.') => {
  const { useState } = globalThis[Symbol.for('interblock:api:hook')]
  assert(useState, `global hook detached`)
  return useState(path)
}
export const usePulse = async (path = '.') => {
  assert.strictEqual(typeof path, 'string')
  assert(path)
  const reply = interchain('@@USE_PULSE', { path })
  return reply
}

export const isApiAction = (request, covenant) => {
  assert.strictEqual(typeof request, 'object')
  assert.strictEqual(typeof request.type, 'string')
  assert(request.type)
  assert.strictEqual(typeof covenant, 'object')
  assert.strictEqual(typeof covenant.api, 'object')
  const schemas = Object.values(covenant.api)
  assert(schemas.length, 'No actions in schema')
  assert(schemas.every((schema) => typeof schema.title === 'string'))
  return schemas.some(({ title }) => request.type === title)
}

export const ensureChild = async (path, installer = 'unity') => {
  assert.strictEqual(typeof path, 'string')
  assert(path)
  // TODO assert child points to a deeper path, not higher one
  if (typeof installer === 'string') {
    installer = { covenant: installer }
  }
  assert.strictEqual(typeof installer, 'object')
  try {
    await interchain(Request.tryPath(path))
    // TODO assert covenant matches
  } catch (error) {
    // TODO work with any child
    const msg = `Segment not present: /.mtab of: .mtab`
    if (error.message !== msg) {
      throw error
    }
    const spawnAction = Request.createSpawn(path, installer)
    await interchain(spawnAction)
  }
}

export const isSystemAction = (request) => {
  assert.strictEqual(typeof request.type, 'string')
  return Request.SYSTEM_TYPES.includes(request.type)
}

export { schemaToFunctions } from './schemaToFunctions'
