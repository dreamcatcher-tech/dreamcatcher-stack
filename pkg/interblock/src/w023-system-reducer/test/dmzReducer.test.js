import { assert } from 'chai/index.mjs'
import { interchain } from '../../w002-api'
import { Request } from '../../w008-ipld'
import { actions } from '..'
import { Engine } from '../../w210-engine'
import Debug from 'debug'
const debug = Debug('interblock:tests:dmzReducer')

describe('dmzReducer', () => {
  describe('spawn', () => {
    test('spawn is implicitly awaited', async () => {
      const reducer = async (request) => {
        debug(`reducer %o`, request)
        if (request.type === 'TEST_SPAWN') {
          await interchain(Request.createSpawn('child1'))
          const result = await interchain('PING', { test: 'ping' }, 'child1')
          debug(`result`, result)
          return { test: true }
        }
      }
      const root = { reducer }
      const engine = await Engine.createCI()
      engine.overload({ root })
      const request = Request.create({ type: 'TEST_SPAWN' })
      // 5 blocks 103ms, 3 blocks 79ms = 20ms / block
      const result = await engine.pierce(request)
      debug(`result`, result)
      const state = engine.latest
      const channel = await state.getNetwork().getChild('child1')
      expect(channel).toMatchSnapshot()
    })
    test('error if no spawn await', async () => {
      // TODO make this error clearer somehow
      const reducer = async (request) => {
        debug(`reducer %o`, request)
        if (request.type === 'TEST_SPAWN') {
          interchain(Request.createSpawn('child1'))
          const result = await interchain('PING', { test: 'ping' }, 'child1')
          debug(`result`, result)
        }
      }
      const root = { reducer }
      const engine = await Engine.createCI()
      engine.overload({ root })
      const request = Request.create({ type: 'TEST_SPAWN' })
      const msg = 'path must be foreign: child1'
      assert.strictEqual(engine.logger.pulseCount, 1)
      await expect(engine.pierce(request)).rejects.toThrow(msg)
      assert.strictEqual(engine.logger.pulseCount, 5)
    })
    test('system ping', async () => {
      const engine = await Engine.createCI()
      const result = await engine.pierce(Request.createPing('test'))
      assert.deepEqual(result, { string: 'test' })
    })
    test.todo('spawn uses hash for seeding inside action')
  })
  describe('openPaths', () => {
    test.todo('cannot be more than one outstanding @@OPEN request in a channel')
    test.todo('missing parents with open children skips straight to children')
    test.todo('3 levels deep')
    test.todo('random path opening and closing and reopening')
    test.todo('simultaneous requests with common parent')
  })
  describe('uplink', () => {
    test.todo('multiple requests from same chain result in single uplink')
  })
  describe('connect', () => {
    test.todo('connect on existing is the same as move')
    test.todo('connect resolves an address without purging queued actions')
    test.todo('connect on existing unknown transmits all queued actions')
    test.todo('connect on operational channel throws')
  })
})
