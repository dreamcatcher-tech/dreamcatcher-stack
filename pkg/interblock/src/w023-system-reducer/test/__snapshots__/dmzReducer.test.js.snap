// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dmzReducer spawn spawn is implicitly awaited 1`] = `
Channel {
  "address": Address {},
  "aliases": Array [
    "child1",
  ],
  "channelId": 3,
  "rx": Rx {
    "reducer": RxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 1,
      "requests": Array [],
      "requestsLength": 0,
    },
    "system": RxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 1,
      "requests": Array [],
      "requestsLength": 0,
    },
    "tip": PulseLink {},
  },
  "tx": Tx {
    "precedent": PulseLink {},
    "reducer": TxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 0,
      "requests": Array [],
      "requestsLength": 1,
    },
    "system": TxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 0,
      "requests": Array [],
      "requestsLength": 1,
    },
  },
}
`;
